lib/bap_primus/bap_primus_generator.ml

```diff
@@ -17,37 +17,45 @@ let generators : Univ_map.t state = Bap_primus_machine.State.declare
     ~uuid:"7e81d5ae-46a2-42ff-918f-96c0c2dc95e3"
     (fun _ -> Univ_map.empty)
 
+module States = Univ_map
 
-module States = Univ_map.With_default
+module type Iter =
+  Iterator.Infinite.S with type dom = int
+
+type 'a iter = (module Iter with type t = 'a)
 
 type 'a gen = {
-  state : 'a States.Key.t;
-  next : 'a -> 'a;
-  value : 'a -> int;
-  min : int;
-  max : int;
+  iter : 'a iter;
+  self : 'a;
+}
+
+type 'a key = 'a gen States.Key.t
+
+type 'a ready = {
+  key : 'a key;
+  gen : 'a gen;
+}
+
+type 'a wait = {
```

Comment:
wait is a bad name here

We basically have two kinds of generators:
- context independent (Ready and Static)
- context dependent

Omitting static for its triviality, the context independent, the context independent has type `'a gen` while context dependent has type `int -> 'a gen`, where `int` denotes the context (it will be the hash of the machine identifier). Both a paired with the key, which is used to store them in the state. We store a value of type `'a gen` in the state. On the first call to the generator (if we don't have the generator stored in the state, then it is the first call), we either take the default state (which we always have) or we use the `init` function to get the context dependent initial value. Then we just apply the generator and store its in the state. On the next calls, if the generator is in the state we always use it, no matter if it is a context dependent or not.